todo_app/app.py

```diff
@@ -2,25 +2,66 @@
 import todo_app.data.session_items as session_items
 from datetime import datetime, timedelta
 from todo_app.view_model import ViewModel
+from flask_login import LoginManager,login_required, UserMixin, login_user, current_user
+from oauthlib.oauth2 import WebApplicationClient
+import os 
+import requests 
+from werkzeug.exceptions import Forbidden
 
+class User(UserMixin):
+    def __init__(self, username):
+        self.id = username
+
+    def writer_role(self): 
+        return self.id == "bob"
+        
 
 def create_app():
     session_items.connect_to_db()
     app = Flask(__name__)
     app.config.from_object("todo_app.flask_config.Config")
 
+    login_manager = LoginManager()
+    client = WebApplicationClient(os.getenv('client_id'))
+    
+    @login_manager.unauthorized_handler 
+    def unauthenticated():
+        redirect_url = client.prepare_request_uri('https://github.com/login/oauth/authorize')
+        return redirect(redirect_url)
+
+    @login_manager.user_loader 
+    def load_user(user_id):
+        return User(user_id)
+    login_manager.init_app(app) 
+
+    @app.route("/login/callback")
+    def login_callback():
+        code = request.args.get('code')
+        url, headers, body = client.prepare_token_request('https://github.com/login/oauth/access_token', client_secret = os.getenv('client_secret'), code = code)
+        response = requests.post(url, headers= headers, data = body)
+        client.parse_request_body_response(response.text)
+        url, headers, body = client.add_token('https://api.github.com/user')
+        response = requests.get(url, headers=headers)
+        user = User(response.json()['login'])
+        login_user(user)
+
+        return redirect("/")
+
+
     @app.route("/")
+    @login_required
     def index():
         items = session_items.get_items()
-        item_view_model = ViewModel(items)
-
+        item_view_model = ViewModel(items, current_user)
         return render_template("index.html", view_model=item_view_model)
 
     @app.route("/create_item", methods=["POST"])
     def create_item():
-        new_item_name = request.form["newitem"]
-        session_items.add_item(new_item_name)
-
+        if current_user.writer_role():
+            new_item_name = request.form["newitem"]
+            session_items.add_item(new_item_name)
+        else:
+            raise Forbidden("You are not allowed to be here!")
```

Comment:
Nice, I wasn't able to avoid this. It would be good to also apply it to the delete/update/change due date methods too since we wouldn't want people messing with those either, and currently non-writers can still delete to-dos if they know what they're doing!
(Similarly every route below should have `@login_required` set, otherwise people can access them without a role at all!)